### DISCUSSION TITLE
Linter for Third Party Packages

### DISCUSSION THREAD

**Discussion by nunnatsa** [golangci/golangci-lint#2969] [Q&A] [answered]
Hi.

I'm working on a linter to ginkgo and gomega (the popular testing package and its matchers packege). The code is here: https://github.com/nunnatsa/ginkgolinter

The linter enforces standards of using ginkgo and gomega. Right now it only warns for asserting `len(x)`, instead of asserting `x` with existing length assertions, such as `HaveLen()`.

I also integrated this new linter with golangci-lint and it's working (locally).

My question is: would you accept a linter for non-pure golang, but for packages, like ginkgo & gomega?

Thanks!

**Comment by ldez** [comment]
Hi,

> would you accept a linter for non-pure golang, but for packages, like ginkgo & gomega?

The question is not precisely that, it's more related to dependencies: more dependencies imply more problems.
Your linter depends on ginkgo and gomega, from my point of view it can be a problem.

You can also create a plugin https://golangci-lint.run/contributing/new-linters/#how-to-add-a-private-linter-to-golangci-lint

**Reply by nunnatsa** [reply]
> Your linter depends on ginkgo and gomega, from my point of view it can be a problem

Thanks @ldez. Good point. The linter itself does not depend on ginkgo and gomega, but only the test file: it won't pass analysis without importing these packages. I wonder if there is a way around it.

**Reply by ldez** [reply]
> I wonder if there is a way around it.

Yes you just have to create a `go.mod` inside `/testdata/src/a`.

For example https://github.com/traefik/mocktail/tree/master/testdata/src/a

You don't need a `vendor` folder.

I think you also have to rethink your test: if your linter can be run without the dependencies, your test must run without the dependency.

**Reply by nunnatsa** [reply]
Thanks. The problem is not the test itself. I can change the test code to use pure go (already changed, thanks for the tip!).

The problem is the go file within `testdata`, in my golangci-lint PR. This one requires ginkgo and gomega, not to run the test - we don't run this file - only analyzing it. But the analyzer will fail without these dependencies.

**Reply by ldez** [reply]
offtopic: you have to remove the `vendor` folder.
https://github.com/nunnatsa/ginkgolinter/tree/main/testdata/src/a/vendor

**Reply by ldez** [reply]
> But the analyzer will fail without these dependencies.

I will see if I can change the test mechanism.

**Reply by nunnatsa** [reply]
I made a small change in my linter - set the Analyzer's `RunDespiteErrors` field to `true` - it makes the analyzer to ignore load errors. However, the cli, that uses `singlechecker.Main`, does not look at this field and fails to run without the ginkgo and gomega dependencies.

The same happens in golangci-lint unit tests, and I couldn't find a way around it, yet.

If there is a way to make golangci-lint respect the Analyzer's `RunDespiteErrors` field, then I think we can contribute linters for 3rd party packages without their dependencies.

**Reply by ldez** [reply]
> If there is a way to make golangci-lint respect the Analyzer's `RunDespiteErrors` field

Currently, it's not possible.

FYI, I'm still working on something.

**Reply by nunnatsa** [reply]
> I'm still working on something

Thanks!

**Comment by nunnatsa** [comment]
> offtopic: you have to remove the vendor folder.

This is to solve a similar problem in the linter unit tests. This  is the only way I found to run the tests.

However, the vendor directory is not part of the package code and  does not affect the package users. It's only used for unit tests.

**Reply by ldez** [reply]
You are right, it's related to `"GO111MODULE=off"` from https://github.com/golang/tools/blob/b929f3bf4d57d9023b392ff552aa21a4845e1a07/go/analysis/analysistest/analysistest.go#L328

**Reply by Antonboom** [reply]
`vendor` unit-tests workaround will be irrelevant after 
https://github.com/golang/go/issues/61336

**Comment by ldez** [comment, marked as answer]
@nunnatsa Good news, I found a solution: https://github.com/golangci/golangci-lint/pull/3124

```
.
├── enabled_linters_test.go
├── fix_test.go
├── linters_test.go
├── output_test.go
├── run_test.go
├── testdata
│   ├── asasalint.go
│   ├── asciicheck.go
│   ├── containedctx.go
...
│   ├── ginkgolinter
│   │   ├── ginkgolinter.go
│   │   ├── go.mod
│   │   └── go.sum
...
│   ├── wrapcheck.go
│   └── wsl.go
...
```

**Reply by nunnatsa** [reply]
Thank you very much @ldez for your efforts.

**Reply by ldez** [reply]
@nunnatsa do you plan to create a PR on golangci-lint?

**Reply by nunnatsa** [reply]
@ldez - I not sure it's belongs to the global linter list. WDYT?

**Reply by ldez** [reply]
I don't understand your sentence, can you rephrase it?

**Reply by nunnatsa** [reply]
OK, here it is: #3369.

@ldez - sorry it took so much time.

**Reply by ldez** [reply]
my login is **L**dez

**Reply by nunnatsa** [reply]
sorry! Thanks for finding that anyway
